### ISSUE TITLE
[BUG] Recent changes on CUDF produces different outputs reading from ORC files.

### ISSUE THREAD

**Issue by rommelDB** [rapidsai/cudf#6600]
**Describe the bug**
Recent changes on CUDF produces different outputs reading from ORC files.

**Steps/Code to reproduce bug**
Reproducer script:
```
filename="/tmp/orders_0_0.orc"
columns=['o_orderkey','o_orderdate']
nrows=10

import cudf
print("cudf")
df = cudf.read_orc(filename, columns=columns)
print(df.head(nrows))
print("")

from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("OrcFileReader").getOrCreate()
df_spark = spark.read.orc(filename).select('o_orderkey','o_orderdate').limit(nrows)
print(spark)
df_spark.show()
print("")

import pandas as pd
print("pandas")
df = pd.read_orc(filename, columns=columns)
print(df.head(nrows))
print("")
```

Sample output:
```
cudf
   o_orderkey         o_orderdate
0           1 1996-01-02 05:00:00
1           2 1996-12-01 05:00:00
2           3 1993-10-14 05:00:00
3           4 1995-10-11 05:00:00
4           5 1994-07-30 05:00:00
5           6 1992-02-21 05:00:00
6           7 1996-01-10 05:00:00
7          32 1995-07-16 05:00:00
8          33 1993-10-27 05:00:00
9          34 1998-07-21 05:00:00

<pyspark.sql.session.SparkSession object at 0x7f8ae3e3d490>
+----------+-------------------+
|o_orderkey|        o_orderdate|
+----------+-------------------+
|         1|1996-01-01 19:00:00|
|         2|1996-11-30 19:00:00|
|         3|1993-10-13 19:00:00|
|         4|1995-10-10 19:00:00|
|         5|1994-07-29 19:00:00|
|         6|1992-02-20 19:00:00|
|         7|1996-01-09 19:00:00|
|        32|1995-07-15 19:00:00|
|        33|1993-10-26 19:00:00|
|        34|1998-07-20 19:00:00|
+----------+-------------------+

pandas
   o_orderkey         o_orderdate
0           1 1996-01-01 19:00:00
1           2 1996-11-30 19:00:00
2           3 1993-10-13 19:00:00
3           4 1995-10-10 19:00:00
4           5 1994-07-29 19:00:00
5           6 1992-02-20 19:00:00
6           7 1996-01-09 19:00:00
7          32 1995-07-15 19:00:00
8          33 1993-10-26 19:00:00
9          34 1998-07-20 19:00:00
```

**Expected behavior**
Cudf output should be the same as other engines.

**Environment overview**
 - Environment location: Bare-metal
 - Method of cuDF install: conda nightly channel, branch-0.17

**Comment by OlivierNV** [comment]
FWIW, the reference version from Apache ORC is [here](https://github.com/apache/orc/blob/master/c%2B%2B/src/Timezone.cc)
I don't think you can just swap the first two entries when building the future rules table (all entries must be swapped afaik if the dst start time is smaller than dst end time), not sure if that's related or if that's a different issue.
It's very tricky to get this covered since cudf writer only writes UTC, and there are lots of funky time zones (it's risky to significantly depart from the reference).

**Comment by kkraus14** [comment]
@vuule I believe you were already looking into ORC timezone related issues so I'm assigning to you, but feel free to kick it to someone else if it's an entirely different issue.

**Comment by vuule** [comment]
> I don't think you can just swap the first two entries when building the future rules table (all entries must be swapped afaik if the dst start time is smaller than dst end time), not sure if that's related or if that's a different issue.
> It's very tricky to get this covered since cudf writer only writes UTC, and there are lots of funky time zones (it's risky to significantly depart from the reference).

I agree, so I have aimed to keep the same behavior in the refactor, but made a few errors that were not covered by our tests.
It's swapping the entries as they are added, not just the first two. 

The root cause of the issue here is that I added 1600 entries instead of 800, and that messes up the binary search when the timestamp is in the period covered by the TZif file.

**Comment by vuule** [comment]
> @vuule I believe you were already looking into ORC timezone related issues so I'm assigning to you, but feel free to kick it to someone else if it's an entirely different issue.

That's fine, I already should have the fix.

**Comment by rommelDB** [comment]
FYI: I performed some tests and the original issue seems to be fixed! Thank you @vuule
